### PR TITLE
Add cmt files to `public_cmi` dir

### LINKED ISSUE #4892: Private modules cause missing build path (in the ocaml-merlin dump)
## The bug / how to reproduce

When a module uses a library that contains private modules, `dune ocaml-merlin --dump-config` shows that the build path doesn’t contain the `byte` directory of the library, but only the `public_cmi` directory. (this causes an error when e.g. trying to locate some definition from this library using merlin)

[merlin_locate_fail.tar.gz](https://github.com/ocaml/dune/files/7108100/merlin_locate_fail.tar.gz) contains a minimal example to reproduce this bug.

Here is the output of `dune build && dune ocaml-merlin --dump-config .` : 

```
A
((STDLIB "C:\cygwin64\home\nchataing\mlfi\mlfi-ins\lib")
 (EXCLUDE_QUERY_DIR)
 (B
  "C:\cygwin64\home\nchataing\merlin_locate_fail\_build/default/.a.eobjs/byte")
 (B
  "C:\cygwin64\home\nchataing\merlin_locate_fail\_build/default/lib/.lib.objs/public_cmi")
 (S "C:\cygwin64\home\nchataing\merlin_locate_fail")
 (S "C:\cygwin64\home\nchataing\merlin_locate_fail\lib")
 (FLG
  (-w
   @1..3@5..28@30..39@43@46..47@49..57@61..62-40
   -strict-sequence
   -strict-formats
   -short-paths
   -keep-locs)))
```

## Spec

- Version of `dune` : 2.9.0
- Version of `ocaml` : 4.12.1
- Operating system: cygwin under windows 10

### PR DESCRIPTION
**Pull request by voodoos**
Fixes #4892 

When private modules are used Dune hides corresponding `cm` files from Merlin by not adding the `bytes` folder to the build path. Exposed modules `cm` files are placed in a special `public_cmi` directory which is added to Merlin's build path.

However this directory was missing `cmt[i]` files causing Merlin's locate command to fail.

This PR enable the symlinking of `cmt[i]`s files into the `public_cmi` directory along with the `cmi` files.

### CODE REVIEW

**Comment by voodoos** [comment]
> Do we already play with symlink in this part of the code? 

Yes, cmi files were already symlinked before when using private modules.

> is it because we don't choose where the compiler put the files?

I am not sure but it looks like the compiler doesn't allow to specify different paths for different artefacts so that might be the case.

File: src/dune_rules/compilation_context.ml [hunk 1]
```diff
@@ -15,7 +15,7 @@ module Includes = struct
            (let+ libs = requires in
             Command.Args.S
               [ iflags libs Byte
-              ; Hidden_deps (Lib_file_deps.deps libs ~groups:[ Cmi ])
+              ; Hidden_deps (Lib_file_deps.deps libs ~groups:[ Cmi; Cmt; Cmti ])
```
Comment (rgrinberg) [inline src/dune_rules/compilation_context.ml:18]: This seems wrong to me. if something depends on `x`, it does not mean we need to build `x`'s cmt's and cmti's. Why are we adding the dependency here?

Comment (voodoos) [reply src/dune_rules/compilation_context.ml:18]: I see, the Cmt's are build as soon as a compilation context has the `bin_annot` field set to true which appear to be the case for every CC used in Dune except for Menhir and when building ppx_driver.

So the cmts are already always built by Dune isn't it ? However I should had a check that the current CC has the bin_annot field set to true.

Comment (rgrinberg) [reply src/dune_rules/compilation_context.ml:18]: Yes, cmt's are essentially always built by dune.

>  However I should add a check that the current CC has the bin_annot field set to true.

I'm not sure why you need this check. Sometimes the cmt's are explicitly turned off for some internally generated modules. That's about the only time when bin_annot is set to false.

**Comment by rgrinberg** [comment]
@voodoos can we revive this PR?

Also, what about an alternative fix for this issue: making the `byte` directory always visible to merlin. Are there any downsides to exposing all the private modules to merlin?

**Comment by rgrinberg** [comment]
The purpose of the public_cmi is to hide it from the user. There's no goal of hiding anything from merlin.

**Comment by rgrinberg** [comment]
Replaced by #6405

**Pull request closed without merging**